### ISSUE TITLE
dnsdist: listen on specific prefix rather than single address

### ISSUE THREAD

**Issue by dmccombs** [PowerDNS/pdns#6832]
- Program: dnsdist
 - Issue type: Feature request

### Short description
This is a feature request that I would intend to implement myself, but want to get a general idea of whether this functionality would be accepted in a pull request before I go forward, so I can weigh other implementation paths I could take before starting.

### Usecase
The ability to listen on a range of addresses easily rather than listening on `0.0.0.0`, to not interfere with other daemons on the same machine that need to listen for traffic on the same port, but on other addresses.

### Description
I would like to add the following functionality:

- The ability to listen on a specific prefix or address range rather than just individual addresses

**Comment by rgacogne** [comment]
> The ability to listen on a prefix rather than just individual addresses

Do you really need to listen on a specific prefix? You can already listen on all available addresses and then restrict via `ACL` or a stateless firewall rule if you need to.

>  Exposing what destination address a query was sent to when processing it in rules and/or Lua

See https://dnsdist.org/reference/dq.html#DNSQuestion.localaddr and https://dnsdist.org/rules-actions.html#NetmaskGroupRule

> Ability to append data to the end of the DNS packet before it's sent to a downstream from a rule and/or Lua

If you mean raw data, outside of any `DNS` standard, I'm not sure we want that. Do you have a specific use case that would require that?

**Comment by pavel-odintsov** [comment]
I like this idea :) We're also looking for some way to listen on prefix because we have thousands of IP addresses to listen on and we can't use 0.0.0.0.

**Comment by dmccombs** [comment]
@rgacogne we have different instances listening on different addresses of the same machine in this case, so being able to listen on a prefix would be needed.

As far as appending data to the end of the packet, our use case is to include some client identifying information for the downstream (which we also control) to be used in deciding an answer to respond with. The incoming request is signed with TSIG, so we don't want to change the DNS data itself (by doing something like using an EDNS option) that would invalidate the signature.

**Comment by Habbie** [comment]
> Ability to append data to the end of the DNS packet before it's sent to a downstream from a rule and/or Lua

Can I strongly suggest that this topic is moved to a separate ticket?

**Comment by dmccombs** [comment]
@rgacogne not yet, I didn't want to start digging into it rather than taking a different path entirely outside of `dnsdist` if the basic ideas weren't likely to be accepted. If they are, we can work to get the implementation to something that's acceptable.

What about the about the concept of appending data to packets?

I'm still not certain this is the direction we'll go in, even if these would likely be accepted if we implemented them, but just want to get an idea of possible paths we could take first.

**Comment by rgacogne** [comment]
> What about the concept of appending data to packets?

We should move that topic to a separate ticket, as suggested by Habbie. Even if are not interested in implementing one without the other, it still makes sense to keep the discussions separate since other people might be interested.

**Comment by dmccombs** [comment]
@rgacogne @Habbie updated this ticket to be specific to listening on a prefix/address range, and opened #6846 to be specific to appending data to packets.

**Comment by rgacogne** [comment]
Thanks!

**Comment by rgacogne** [comment]
Closing this until someone mentions an actual way of listening on prefix/addresses range :)
